### ISSUE TITLE
Granite Speech perf follow-ups from PR #45

### ISSUE THREAD

**Issue by christopherthompson81** [christopherthompson81/vernacula#46]
Follow-up items identified during PR #45 review that weren't blocking for the Granite perf landing but are worth tracking.

## 1. `BindOutput(name, preAllocatedOrtValue)` audit across ASR backends

PR #45 phase 8 measured a ~24% wall-time improvement on Granite from switching the step-loop logits binding from `BindOutputToDevice(\"logits\", cpuMemInfo)` (ORT allocates a fresh CPU buffer per call) to `BindOutput(\"logits\", preAllocatedOrtValue)` (one CPU buffer reused).

Survey of the existing backends:

| Backend | Sites still using `BindOutputToDevice` for logits |
|---|---|
| `WhisperTurbo.cs` | lines 531, 614, 755, 816 |
| `Qwen3Asr.cs` | lines 1651, 1689, 1908, 1937 (line 665 already uses the right pattern) |
| `VibeVoiceAsr.cs` | line 556 |
| `CohereTranscribe.cs` | lines 353, 458 |

Likely 5–10% per-backend wall-time win on step-heavy workloads. Per-backend: write the change, regression-test parity against an existing golden transcript, benchmark. 1–2 days each.

The Qwen3Asr inconsistency (one site already uses the right pattern) is the strongest signal that this is a generally-applicable refactor.

## 2. Extract `mel/encoder/projector` preamble in `GraniteSpeech.cs`

After PR #45 there are 5 separate `Transcribe*` methods (`TranscribeBatch`, `TranscribeBatchStatic`, `TranscribeStaticGqaSingle`, `TranscribeStaticGqaBatched`, `TranscribeDynamicGqaSingle`) that each repeat ~30–40 lines of:

```csharp
var audioT = new DenseTensor<float>(waveform, [1, waveform.Length]);
using var melResults = _mel.Run([NamedOnnxValue.CreateFromTensor(\"audio\", audioT)]);
// ... encoder.Run ... projector.Run ...
```

Extract into one helper like `(float[] audioEmbeds, int audioTokens, long audioDim) RunAudioPipeline(float[] waveform, ref long melMs, ref long encMs, ref long projMs)`. Saves ~200 lines and removes a class of copy-paste drift.

## 3. `tests/GraniteSpeechSmoke` coverage for new bundle variants

PR #45 lands auto-detection across multiple bundle types (`_isStaticBundle`, `_isGqaBundle`, `_hasPositionIdsInput`, `_isDynamicGqaBundle`, `_hasNextTokenOutput`). The existing smoke test runs against one bundle layout (the original FP32 dynamic). It would have caught two real bugs we hit during PR #45 development:

- Phase 8: output-binding index mismatch (interleaved `key_0, value_0, key_1, …` bind vs grouped `[all keys, then all values]` indexing) produced plausible-byte-but-garbage transcripts.
- Phase 8: `using var outputs = stepBinding.GetOutputValues()` disposed inner OrtValues, leaving `pastKvs` with dangling references → infinite loop.

Add one smoke run per bundle variant: dynamic-MHA (recommended), static-bf16 (back-compat), dynamic-GQA, optionally `--add-argmax` variant. ~1 day.

## 4. Static-shape `Transcribe*` cleanup

PR #45 left `TranscribeBatchStatic`, `TranscribeStaticGqaSingle`, `TranscribeStaticGqaBatched` in place with docstrings noting their superseded status (Run 20 phase 4 onwards). After a release cycle confirms no consumers depend on static-shape bundles, these three methods plus the `_staticZeroPastKv` ctor plumbing (lines 184–214) can be deleted (~700 lines).

## Context

PR: #45
Doc: `docs/dev/granite_speech_perf_investigation.md` Runs 18–20

**Comment by christopherthompson81** [comment]
Worked through this. PR #137 lands item 2; items 1, 3 and 4 need decisions rather than code.

### Item 1 — measured, does not reproduce

Full log: `docs/dev/bind_output_logits_investigation.md`.

I wrote the change for `WhisperTurbo.TranscribeBatch` and benchmarked it warm in three
workload shapes, including the one that maximises the effect (batched *and* step-heavy,
1.6 MB of per-step allocation removed):

| workload | steps | step loop share of phases | baseline | preallocated |
|---|---|---|---|---|
| VAD-segmented, B=8 | 57 | 19 % | 125–128 ms | 128–129 ms |
| gapless, B=1 | 140 | 29 % | 152–155 ms | 152–153 ms |
| gapless ×8, B=8 | 393 | 33 % | 797–800 ms | 797–798 ms |

No win anywhere, transcripts byte-identical. At ~2 ms of GPU work per step, ORT's CPU
allocator serving the logits buffer isn't measurable — the step is bound by decoder
compute.

The site survey also shrinks the remaining scope:

- **`Qwen3Asr:665` is the default path and already uses `BindOutput`.** The other two
  step-loop sites (`DecodeOnGpuUnified`, `DecodeWithIoBinding`) only run under
  `--qwen3asr-serial` or with a split-decoder bundle. The "inconsistency" the issue read
  as a signal is actually the hot path having already been done.
- **`VibeVoiceAsr:556` isn't the same kind of site.** It's in `RunOnce`, shared between
  prefill and decode, with a logits shape that differs between the two — so a single
  preallocated buffer needs worst-case sizing and re-wrapping. Different, larger change.
  No VibeVoice bundle here to regression-test it.
- **The init/prefill sites aren't candidates at all** (Whisper 531/755, Qwen3 1651/1908,
  Cohere 353). They run once per segment, not once per token, and their `[B, S, V]` shape
  varies per segment. Nothing to amortise.

On why PR #45 saw 24 % and this sees 0 %: I couldn't run the decisive experiment. Granite's
two `BindOutput(preallocated)` sites are in `TranscribeBatch` (dynamic-MHA) and
`TranscribeStaticGqaBatched` (static-GQA); the only bundle I have is static-MHA, which
dispatches to `TranscribeBatchStatic` — no IO binding at all. My guess is the 24 % wasn't
the allocation but the `next_token` GPU-argmax output added in the same phase-8 commit,
which removes an entire `[B, 1, 49k]` device→host copy per step. That's a much larger
effect than allocating the buffer the copy lands in, and it'd be the thing worth chasing
for the other backends.

**Suggest closing item 1 as measured-and-declined**, or narrowing it to "can the other
decoders be re-exported with a `next_token` output".

### Item 3 — blocked on bundles, not on effort

Needs dynamic-MHA, dynamic-GQA and static-GQA bundles to run one smoke per variant. This
machine has static-MHA only. Worth keeping open; it's the reason four of the five call
sites PR #137 touches are compiler-verified rather than executed.

### Item 4 — please don't do this one

> After a release cycle confirms no consumers depend on static-shape bundles, these three
> methods … can be deleted (~700 lines).

There is a consumer: **the only working Granite bundle I have is static-shape**, and it
routes through `TranscribeBatchStatic` — one of the three methods slated for deletion.
`Recognize` → `TranscribeBatch` → `if (_isStaticBundle) return TranscribeBatchStatic(…)`.

The recommended dynamic-MHA export is also not simply available: the newest Granite export
here (`granite_speech_4_1_2b_static_mha_bf16`) fails at session creation with

```
Type Error: Type parameter (T) of Optype (MultiHeadAttention) bound to different types
(tensor(bfloat16) and tensor(float)) in node (mha_l0_node)
```

so the MHA-fused export path is itself producing broken graphs right now. Deleting the
static-shape methods would remove the only Granite path that currently runs, in favour of
one that currently doesn't. Suggest item 4 stays blocked until item 3 exists and a
dynamic-MHA bundle is confirmed loadable.
